server/src/main/java/org/opensearch/index/engine/IngestionEngine.java

```diff
@@ -0,0 +1,1016 @@
+                new TranslogStats(0, 0, 0, 0, 0),
+                new Translog.Snapshot() {
+                    @Override
+                    public void close() {}
+
+                    @Override
+                    public int totalOperations() {
+                        return 0;
+                    }
+
+                    @Override
+                    public Translog.Operation next() {
+                        return null;
+                    }
+                }
+            );
+            documentMapperForType = engineConfig.getDocumentMapperForTypeSupplier().get();
+
+            IngestionSource ingestionSource = Objects.requireNonNull(indexMetadata.getIngestionSource());
+            ingestionConsumerFactory = Objects.requireNonNull(engineConfig.getIngestionConsumerFactory());
+            // initialize the ingestion consumer factory
+            ingestionConsumerFactory.initialize(ingestionSource.params());
+            String clientId = engineConfig.getIndexSettings().getNodeName()
+                + "-"
+                + engineConfig.getIndexSettings().getIndex().getName()
+                + "-"
+                + engineConfig.getShardId().getId();
+            IngestionShardConsumer ingestionShardConsumer = ingestionConsumerFactory.createShardConsumer(
+                clientId,
+                engineConfig.getShardId().getId()
+            );
+            logger.info("created ingestion consumer for shard [{}]", engineConfig.getShardId());
+
+            Map<String, String> commitData = commitDataAsMap();
+            StreamPoller.ResetState resetState = StreamPoller.ResetState.valueOf(
+                ingestionSource.getPointerInitReset().toUpperCase(Locale.ROOT)
+            );
+            IngestionShardPointer startPointer = null;
+            Set<IngestionShardPointer> persistedPointers = new HashSet<>();
+            if (commitData.containsKey(StreamPoller.BATCH_START)) {
+                // try recovering from commit data
+                String batchStartStr = commitData.get(StreamPoller.BATCH_START);
+                startPointer = ingestionConsumerFactory.parsePointerFromString(batchStartStr);
+                try (Searcher searcher = acquireSearcher("restore_offset", SearcherScope.INTERNAL)) {
+                    persistedPointers = fetchPersistedOffsets(Lucene.wrapAllDocsLive(searcher.getDirectoryReader()), startPointer);
+                    logger.info("recovered persisted pointers: {}", persistedPointers);
+                } catch (IOException e) {
+                    throw new EngineCreationFailureException(config().getShardId(), "failed to restore offset", e);
+                }
+                // reset to none so the poller will poll from the startPointer
+                resetState = StreamPoller.ResetState.NONE;
+            }
+
+            streamPoller = new DefaultStreamPoller(startPointer, persistedPointers, ingestionShardConsumer, this, resetState);
+            streamPoller.start();
```

Comment:
The pattern here of sending the `this` pointer out of the constructor is a bit concerning to me. This allows DefaultStreamPoller to observe a partially-constructed instance of IngestionEngine. If, for example, a future refactoring added a new `final` field to this class and it were initialized after this line, then DefaultStreamPoller would be able to observe an uninitialized final field. Alternatively, if `streamPoller.start()` throws an exception then that `streamPoller` instance will still have a reference to this instance, which never successfully completed its constructor. Is there a way to structure these classes to avoid these problems?

Reply:
I mainly need it to pass the engine to `MessageProcessor`, because `MessageProcessor` can invoke the index/delete operations on the engine. 

you have a valid point, and ideally we can call `engine.start`, which can pass the pointer to the poller to start. do you feel it's viable with the existing Engine interface design?

Reply:
Would it help to do this in the IngestionEngineFactory? Something like:

```java
@Override
public Engine newReadWriteEngine(EngineConfig config) {
    final IngestionEngine engine = new IngestionEngine(config);
    engine.start();
    return engine;
}
```

By the way, this pattern of having complex logic inside constructors (and potentially leaking references to partially-constructed instances) is pervasive throughout this code base so I'm not surprised the class design resulted in this. I'll get off my soapbox now :)

Reply:
nice. yes, it works